### PR TITLE
Add &numberify option to to-json builtin

### PR DESCRIPTION
**Pull request by VictorLowther**
I have a usecase that involves taking JSON from a remote source,
modifying one or more fields in it, and updating the remote source
with the new JSON fields.

Elvish internally translates all the numeric fields in the recieved
JSON into strings, which causes the JSON I need to send back to be
invalid and rejected by the remote service.  Until elvish gains
support for ints and floats as native internal data types, &numberify
is a hack to turn any strings that parse as valid ints or floats into
actual ints or floats.

### CODE REVIEW

File: eval/builtin_fn_io_test.go [hunk 1]
```diff
@@ -19,17 +15,19 @@ func TestBuiltinFnIO(t *testing.T) {
 		That(`print "a\nb" | slurp`).Puts("a\nb"),
 		That(`print "a\nb" | from-lines`).Puts("a", "b"),
 		That(`print "a\nb\n" | from-lines`).Puts("a", "b"),
-		That(`echo '{"k": "v", "a": [1, 2]}' '"foo"' | from-json`).Puts(
```
Comment (xiaq) [inline eval/builtin_fn_io_test.go]: Did you remove this test case by mistake?

File: eval/builtin_ns.go [hunk 2]
```diff
@@ -13,6 +14,7 @@ var builtinNs = Ns{
 	"ok":    vars.NewRo(OK),
 	"true":  vars.NewRo(true),
 	"false": vars.NewRo(false),
+	"nil":   vars.NewRo(vals.Nil{}),
```
Comment (xiaq) [inline eval/builtin_ns.go]: I don't want to introduce a nil value for Elvish. Call this `-json-nil`.

File: eval/builtin_fn_io.go [hunk 3]
```diff
@@ -122,13 +125,48 @@ func toLines(fm *Frame, inputs Inputs) {
...
+		return res
+	default:
+		return src
+	}
+}
+
 // toJSON converts a stream of Value's to JSON data.
-func toJSON(fm *Frame, inputs Inputs) {
+func toJSON(fm *Frame, rawOpts RawOptions, inputs Inputs) {
+	opts := struct{ Numberify bool }{false}
```
Comment (xiaq) [inline eval/builtin_fn_io.go]: Call the option `&-numberify` by using a field tag `` `name="-numberify"` ``. Leading dashes mark things that will likely go away.

File: eval/vals/nil.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package vals
+
+// Represents the nil value, without actually being nil
+// which otherwise causes problems doing equality testing.
+type Nil struct{}
```
Comment (xiaq) [inline eval/vals/nil.go]: Call this `JsonNil` and document that it is only used for JSON interop.

File: eval/vals/repr.go [hunk 5]
```diff
@@ -32,6 +32,8 @@ type Reprer interface {
 // format "<unknown %v>".
 func Repr(v interface{}, indent int) string {
 	switch v := v.(type) {
+	case Nil:
```
Comment (xiaq) [inline eval/vals/repr.go]: This is not a builtin type and thus doesn't need special-casing. Implement `Reprer` for `Nil` (or rather `JsonNil`, the name I suggested).

File: runtime/runtime.go [hunk 6]
```diff
@@ -22,6 +22,7 @@ import (
 )
 
 var logger = util.GetLogger("[runtime] ")
+var ExtraNS = map[string]eval.Ns{}
```
Comment (xiaq) [inline runtime/runtime.go]: This is unrelated.

File: runtime/runtime.go [hunk 7]
```diff
@@ -97,6 +98,9 @@ func InitRuntime(binpath, sockpath, dbpath string) (*eval.Evaler, string) {
 		ev.InstallModule("store", storemod.Ns(client))
 		ev.InstallModule("daemon", daemonmod.Ns(client, spawner))
 	}
+	for k, v := range ExtraNS {
```
Comment (xiaq) [inline runtime/runtime.go]: Ditto.

File: eval/builtin_fn_io.go [hunk 8]
```diff
@@ -122,13 +125,48 @@ func toLines(fm *Frame, inputs Inputs) {
 	})
 }
 
+// Exported so that other packages can refer to it for proper nil
```
Comment (xiaq) [inline eval/builtin_fn_io.go]: * This doesn't need to be exported; it is not used in another other package.
* This function is only needed when "numberify" is enabled. Just name it `numberify` and avoid calling it when the option is not given.
* Instead of only handling marshaling of `Nil` here, implement [`Json.Marshaler`](https://godoc.org/encoding/json#Marshaler) for the `Nil` type so that it can be marshaled into JSON using the standard JSON encoding functions.

File: eval/builtin_fn_io_test.go (hunk 1 above)
Comment (VictorLowther) [reply eval/builtin_fn_io_test.go]: No, the test case I added made it redundant.

File: eval/builtin_ns.go (hunk 2 above)
Comment (VictorLowther) [reply eval/builtin_ns.go]: Can do.

File: eval/builtin_fn_io.go (hunk 3 above)
Comment (VictorLowther) [reply eval/builtin_fn_io.go]: OK, will do.

File: eval/vals/repr.go (hunk 5 above)
Comment (VictorLowther) [reply eval/vals/repr.go]: Will do.

File: runtime/runtime.go (hunk 6 above)
Comment (VictorLowther) [reply runtime/runtime.go]: You want me to pull these into a seperate PR?

File: eval/builtin_fn_io.go (hunk 8 above)
Comment (VictorLowther) [reply eval/builtin_fn_io.go]: I call it from Go functions that are implemented in a namespace I inject using the ExtraNS stuff -- that namespace is responsible for implementing in-shell access to a JSON baseed API that does not deal well with numbers and nulls magically transmuting to strings.  If this is not a good place for it, is there a better location?

File: eval/builtin_fn_io_test.go (hunk 1 above)
Comment (xiaq) [reply eval/builtin_fn_io_test.go]: I don't see what you mean; you haven't added a test case against the output of `from-json`.

File: eval/builtin_fn_io.go (hunk 8 above)
Comment (xiaq) [reply eval/builtin_fn_io.go]: Elvish does not have a stable Go API. Helper functions like these are especially subject to change. If you want to use Elvish from another project, it is a better idea to duplicate those code in the consumer package. The other APIs you use are also subject to change, but less so.

File: runtime/runtime.go (hunk 6 above)
Comment (xiaq) [reply runtime/runtime.go]: Please see my comment above about Go API; if you are building an alternative runtime, consider duplicating some code from the runtime package.

File: eval/builtin_fn_io_test.go (hunk 1 above)
Comment (VictorLowther) [reply eval/builtin_fn_io_test.go]: Does the test case that tests that going from json, then back to the same json not cover it?

File: runtime/runtime.go (hunk 6 above)
Comment (VictorLowther) [reply runtime/runtime.go]: My intent is not to build an alternative runtime, it is to make it easy to add new namespaces that are implemented in go. Doing that in a way that does not involve just forking elvish means I need to at least be able to have the runtime register new go foncs in their own namespace.

Comment (xiaq) [reply runtime/runtime.go]: Sorry if I wasn't clear.

Where do you intend to add new namespaces implemented in Go?

* Are you going to add them in the Elvish codebase? Then the current approach of calling `ev.InstallModule` works well.

* Are going to add them outside the Elvish codebase? Then that is what I meant by "an alternative runtime". Doing so means that you will be inevitably relying on Elvish's Go API. I am not ready to commit to a stable Go API, so I don't intend to support such use cases.

File: eval/builtin_fn_io_test.go (hunk 1 above)
Comment (xiaq) [reply eval/builtin_fn_io_test.go]: No. That only tests that `from-json` and `to-json` are inverse operations (under certain conditions). If you both `from-json` and `to-json` are simply identity functions, the test you added would pass.

File: eval/builtin_ns.go (hunk 2 above)
Comment (xiaq) [reply eval/builtin_ns.go]: Sorry for the late feedback, but now that this is only used for JSON, `json-null` is actually a more sensible choice...

**Comment by xiaq** [comment]
Thanks for your interest and contribution, but this is no longer necessary as we have a design for explicitly number types in #816.
